### PR TITLE
Add support of DDP and experimental CompiledAutograd

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #432

Summary:
Address the comments in https://github.com/pytorch/torchtitan/pull/319 and resubmit the PR to fit the current code base.

Test Plan:
```
CONFIG_FILE=./train_configs/debug_model.toml ./run_llama_train.sh --comm.train_timeout_seconds=3600   --training.tensor_parallel_degree=1 --training.data_parallel_degree=8 --experimental.data_parallel_type=ddp --training.steps=1000 --metrics.log_freq=10 --profiling.profile_freq=1000
```

**DDP tensorboard log**
Screenshot 2024-07-18 at 9 42 43 AM
Screenshot 2024-07-18 at 9 43 03 AM

**FSDP tensorboard log**
Screenshot 2024-07-18 at 9 44 36 AM
Screenshot 2024-07-18 at 9 45 46 AM

### CODE REVIEW

**Comment by tianyu-l** [comment]
Looks better to me! Shall we add an integration test or a set of them in `./test_runner.py` for this feature?

**Comment by fegin** [comment]
> Looks better to me! Shall we add an integration test or a set of them in `./test_runner.py` for this feature?

I added the test but it seems that the 4/8 GPUs integration tests are broken?

**Comment by tianyu-l** [comment]
> I added the test but it seems that the 4/8 GPUs integration tests are broken?

The 4 GPUs integration test fails because we need to adapt the changes in `train.py` in `estimate.py` as well.
For details pls see https://github.com/pytorch/torchtitan/pull/425#issuecomment-2189788501

The 8 GPUs test fails is irrelevant. Please ignore it for now.

**Comment by fegin** [comment]
@tianyu-l I fixed the 4 GPU issue. Does the PR look good to land?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 1]
```diff
@@ -489,6 +490,24 @@ def apply_dp(model, world_mesh, parallel_dims, job_config: JobConfig):
     return model
 
 
+def apply_ddp(model, world_mesh, parallel_dims, job_config: JobConfig):
+    if world_mesh.ndim > 1:
+        raise RuntimeError("DDP has not supported > 1D parallelism.")
+
+    if job_config.training.compile:
+        if job_config.experimental.compiled_autograd:
```
Comment (tianyu-l) [inline torchtitan/parallelisms/parallelize_llama.py]: `compiled_autograd` -> `enable_compiled_autograd`?

File: test_runner.py [hunk 2]
```diff
@@ -273,6 +273,19 @@ def build_test_list():
             "fsdp2_mem_tracker",
             ngpu=4,
         ),
+        OverrideDefinitions(
+            [
+                [
+                    "--training.tensor_parallel_degree 1",
+                    "--training.data_parallel_degree 8",
```
Comment (tianyu-l) [inline test_runner.py]: I think we don't need to specify this, just using default is ok.

File: test_runner.py [hunk 3]
```diff
@@ -273,6 +273,19 @@ def build_test_list():
...
+                    "--experimental.data_parallel_type ddp",
+                    "--experimental.enable_compiled_autograd",
+                ]
+            ],
+            "CompiledDDP",
+            "compiled_ddp",
+            ngpu=8,
```
Comment (tianyu-l) [inline test_runner.py]: why do we need to use 8 gpus? A side note is that it doesn't mean it passes the 4 gpu test.

File: estimation.py [hunk 4]
```diff
@@ -67,6 +67,7 @@ def estimate_memory(job_config: JobConfig):
         pp=job_config.experimental.pipeline_parallel_degree,
         world_size=world_size,
         enable_loss_parallel=job_config.training.enable_loss_parallel,
+        dp_type=job_config.experimental.data_parallel_type,
```
Comment (tianyu-l) [inline estimation.py]: I'm not sure what's the best category here. I think FSDP is not experimental. Is DDP without compiled_autograd also experimental? I'm OK with leaving this option as experimental if things are not clear.

**Review by tianyu-l — COMMENTED** [review]
I wonder if we should do some perf tests and even loss convergence tests before merging. If it is well-known that DDP works very well, then it may be fine to just land. But even so, it's good to publish some numbers and curves in the PR summary, e.g. compared with 1D FSDP2.

**Comment by fegin** [comment]
@tianyu-l Agree we should test the accuracy as that's also one reason we want to bring in DDP. But what's the meaning of performance? We can only test small models with DDP or trim the batch size and sequence length due to the memory usage. The performance may not be good with any parallelisms as we cannot saturate the GPU. And bf16 is not enabled for DDP.

**Comment by tianyu-l** [comment]
> The performance may not be good with any parallelisms as we cannot saturate the GPU.

Do you mean the comparison with other parallelism (e.g. FSDP2) is not fair on small models even under the same setup? I was wondering if we can have the following experimental setup to verify performance:
1. DDP, customized model (somewhere between debug model and 8B model as long as it can fit), fp32 precision, 8 GPUs
2. FSDP2, same model, same precision, same number of GPUs

In this case, should we expect DDP to have similar or higher WPS than FSDP2? (without all-gather in fwd/bwd, but reduce-scatter becomes all-reduce in bwd)

**Comment by fegin** [comment]
@tianyu-l No, I meant even if we use the same configuration, the performance bottleneck is likely not in the communication, so the performance improvement is very little. I tested with the debug model. The performance of DDP and FSDP are almost the same, DDP may be slightly higher but the improvement is too little to make any conclusion. 8B model is going to be OOM.

File: torchtitan/parallelisms/parallelize_llama.py (hunk 1 above)
Comment (fegin) [reply torchtitan/parallelisms/parallelize_llama.py]: oops, my bad, updated

**Review by tianyu-l — APPROVED** [review]
After some offline discussion, I believe the PR is ready to land. Please post some numerics / perf numbers before merge.

Note: DDP is widely-believed to be "correct" and can serve as a foundation to verify numerics / loss convergence for new technologies. However, due to it's memory constraints, there isn't a verified toml config with proper hyperparameters (e.g. learning rate, batch size). For now let's still rely on 1D FSDP2 as the test base.
